packages/protocol/contracts/liquidity/GrandaMento.sol

```diff
@@ -0,0 +1,477 @@
+    require(
+      sellToken.transferFrom(msg.sender, address(this), sellAmount),
+      "Transfer in of sell token failed"
+    );
+
+    // Record the proposal.
+    // Add 1 to the running proposal count, and use the updated proposal count as
+    // the proposal ID. Proposal IDs intentionally start at 1 rather than 0 because
+    // LinkedList, which is used for keeping track of active proposal IDs, requires
+    // keys to be non-zero.
+    exchangeProposalCount = exchangeProposalCount.add(1);
+    exchangeProposals[exchangeProposalCount] = ExchangeProposal({
+      exchanger: msg.sender,
+      stableToken: stableToken, // for stable tokens, is saved in units to deal with demurrage.
+      sellAmount: sellCelo ? sellAmount : IStableToken(stableToken).valueToUnits(sellAmount),
+      buyAmount: buyAmount,
+      approvalTimestamp: 0, // initial value when not approved yet
+      state: ExchangeProposalState.Proposed,
+      sellCelo: sellCelo
+    });
+    // Push it into the array of active proposals.
+    activeProposalIds.push(exchangeProposalCount);
+    // Even if stable tokens are being sold, the sellAmount emitted is the "value."
+    emit ExchangeProposalCreated(
+      exchangeProposalCount,
+      msg.sender,
+      stableToken,
+      sellAmount,
+      buyAmount,
+      sellCelo
+    );
+    return exchangeProposalCount;
+  }
+
+  /**
+   * @notice Approves an existing exchange proposal.
+   * @dev Sender must be the approver. Exchange proposal must be in the Proposed state.
+   * @param proposalId The identifier of the proposal to approve.
+   */
+  function approveExchangeProposal(uint256 proposalId) external onlyApprover {
+    ExchangeProposal storage proposal = exchangeProposals[proposalId];
+    // Ensure the proposal is in the Proposed state.
+    require(proposal.state == ExchangeProposalState.Proposed, "Proposal must be in Proposed state");
+    // Set the time the approval occurred and change the state.
+    proposal.approvalTimestamp = block.timestamp;
+    proposal.state = ExchangeProposalState.Approved;
+    emit ExchangeProposalApproved(proposalId);
+  }
+
+  /**
+   * @notice Cancels an exchange proposal.
+   * @dev Only callable by the exchanger if the proposal is in the Proposed state
+   * or the owner if the proposal is in the Approved state.
+   * @param proposalId The identifier of the proposal to cancel.
+   */
+  function cancelExchangeProposal(uint256 proposalId) external nonReentrant {
+    ExchangeProposal storage proposal = exchangeProposals[proposalId];
+    // Require the appropriate state and sender.
+    // This will also revert if a proposalId is given that does not correspond
+    // to a previously created exchange proposal.
+    require(
+      (proposal.state == ExchangeProposalState.Proposed && proposal.exchanger == msg.sender) ||
+        (proposal.state == ExchangeProposalState.Approved && isOwner()),
+      "Sender cannot cancel the exchange proposal"
+    );
```

Comment:
Hmm that's a good question. Part of the reasoning why only Governance can cancel in the `Approved` state was to prevent the exchanger from having a "free option" where they can just cancel the exchange when the price they locked in at proposal-time is worse than the price at exchange-execution-time. Giving the approver the ability to cancel when it's Approved could make it easier for the exchanger to do this if they're able to convince the approver to cancel it on their behalf. But allowing this could prevent needing to rally the community behind a governance proposal if the approvers mistakenly approved a proposal. I'm generally supportive of it, what're your thoughts?

Reply:
Following the brief discussion [here](https://github.com/celo-org/celo-proposals/issues/229#issuecomment-866414656), deciding to keep as is